README.txt
-----------------------------------------------------------------------------------
Project: OpenGTS - Open GPS Tracking System
URL    : http://www.opengts.org
File   : README.txt
-----------------------------------------------------------------------------------

Notes:
 - The following DCS modules are still present in the package, but have been removed
   from the build "all" ant target:
      icare, sipgear, aspicore, gtsdmtp, taip, gc101, mologogo, gpsmapper
   These modules can be built separately as follows, should you need to use them:
      ant icare sipgear aspicore gtsdmtp taip gc101 mologogo gpsmapper
 - Support for the CelltracGTS/Server requires downloading/installing the library
   support for the "ctrac.war" servlet. Please see the following for more information:
      http://www.geotelematic.com/CelltracGTS
 - Please refer to the included "OpenGTS_Config.pdf" file for all installation and 
   configuration information for the OpenGTS system.
 - When updating to a newer version of OpenGTS, it is highly recommended that the 
   following command be run to update the database with the latest table/column 
   changes:
     $  bin/dbAdmin.pl -tables=ca
   Or, run the following if alternate key fields also need to be updated:
     $  bin/dbAdmin.pl -tables=cak
 - On Windows 7, Internet Explorer 8, if you receive the error 
      Error [jsMapInit:[object error]
   then you may need to enable "Active Scripting" in the security tab.
 - A listing of the database tables and fields can be found in the file "SCHEMA.txt"
   in the OpenGTS installation diretory.
 - Documentation is available online at the following link:
     http://www.opengts.org/documentation.html
 - This open-source version of OpenGTS includes support for the following GPS 
   tracking devices:
     - Aspicore GSM Tracker (Nokia, Samsung, Sony Ericsson phones).
       [http://www.aspicore.com]
     - Astra Telematics AT240, AT110, AT210.
       [http://www.gps-telematics.co.uk]
     - Sanav GC-101, MT-101, and CT-24 Personal Tracker (HTTP-based protocol).
       Sanav GX-101 Vehicle Tracker (HTTP-based protocol).
       [http://www.sanav.com]
     - V-Sun 3338 Personal Tracker.
     - CelltracGTS/Free Android application
     - CelltracGTS/Pro Android application 
     - Most TK102/TK103 devices.
     - ZhongShan SIPGEAR Technology Co, Ltd.
     - TAIP (Trimble ASCII Interface Protocol).
     - TrackStick GPS data logger.
     - "GPSMapper" capable phones.
     - "NetGPS" capable devices.

Please let us know if you find any issues with this release.

-----------------------------------------------------------------------------------

Visit the following link for additional downloadable documentation and add-ons for 
OpenGTS:
  http://www.opengts.org/documentation.html
  http://www.geotelematic.com/documentation.html

-----------------------------------------------------------------------------------

This package may contain one or more of the following source modules which contain 
copyrights from their respective authors.  Please review these source modules for 
their copyright and license information:
   war/track/js/mapstraction/mapstraction.js
   war/track/js/mapstraction/mapstraction-geocode.js
   war/track/js/mapstraction/labeledmarker.js
   war/track/js/sorttable/sorttable.js

-----------------------------------------------------------------------------------

Contributors:
  See the following link for a list of contributors:
    http://www.opengts.org/info.html

-----------------------------------------------------------------------------------

Sample data:
Additional documentation for installing sample 'demo' data into the database can be
found in the "README.txt" file in the "sampleData" directory at
    sampleData/README.txt

-----------------------------------------------------------------------------------

Sanav GC-101 GPS Tracking Device Support:
Additional documentation for installing and configuring the GC-101 server within
OpenGTS can be found in the "README.txt" file in the "gc101" source directory at
   src/org/opengts/war/gc101/README.txt

See the following link for manufacturer's product information:
   http://www.sanav.com

-----------------------------------------------------------------------------------

TAIP (Trimble ASCII Interface Protocol) Support:
Support for TAIP (Trimble ASCII Interface Protocol) is included in this release.  
This server uses the raw-socket mode device communication server based on the example
'template' server.  The "taip.jar" server jar file can be built with the command:
   > ant taip
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s taip

-----------------------------------------------------------------------------------

Updating existing Geozone table entries 
(only required when upgrading from versions previous to v1.9.3):

If you have existing Geozone table entries, they will need to be updated to include
Bounding-Box information used to optimize Geozone lookups.  This update process is
done automatically with the following steps:

1) Update the new Geozone bounding-box table fields:
     > bin/dbAdmin.pl -tables=ca

2) For each account having Geozone entries, run the following command (replace "<accountID>" 
   with the appropriate account-IDs for each account which has Geozone entries):
     > bin/admin.pl Geozone -account=<accountID> -list -update

The update process should now be complete.

-----------------------------------------------------------------------------------

Runtime config file property key references
(for *.conf files, such as 'default.conf', 'common.conf', 'webapp.conf, ...):
- The prefixing "#" is a comment character, indicating that the following text is for
  informational purposes only, and should not be parsed.  Comments must be placed on
  a separate line, and must not be specified at the end of a property specification.
- Previously defined property keys may be referenced by placing them in ${...} brackets,
  such as ${log.name}.  Environment variables may also be referenced in this manner, as in 
  ${GTS_HOME}, or ${CATALINA_HOME}.
- A default value may be specified for referenced property keys which have not been defined.   
  For instance ${OUTPUT_DIR=/tmp} would resolve to "/tmp", if the property OUTPUT_DIR is not 
  defined (as an environment variable, or otherwise), or will resolve to the value specified 
  on the OUTPUT_DIR property (or environment variable) if it is defined.
- Recursive propery key references may also be specified.  For instance, the specification
  ${THIS_DIR=${THAT_DIR=/tmp}} will first attempt to resolve the value for "THIS_DIR", if not
  found, then value of THAT_DIR will be returned.  If THAT_DIR is not defined, then finally
  the specification will resolve to "/tmp".
- Property reference specifications for which the property is not defined, and there is no
  specified default value will resolve to the literal "${var}" string.  That is, if "var"
  is not defined, then ${var} will resolve to the literal string "${var}".
- Property reference specifications are only resolved at the time a property value is 
  requested, not at the time the property key=value line is parsed (ie. lazy resolution).  
  For instance, in the property specification below, "var_a" will resolve to "test", even 
  though "var_b" is not defined until after "var_a":
    var_a=${var_b}
    var_b=test
  This means that if "var_b" should ever change, the resolved value of "var_a" will also
  change accordingly.  [ie. "var_a" will only be resolved at the time a call to an RTConfig
  property value retrieval method is called - such as RTConfig.getString("var_a")].
  Write-only (assignable) property keys (ie. %log, %include) are the only exception to this 
  rule.  Since their assigned behavior is executed at the time they are parsed ('%log' prints
  a log message, and '%include' includes a file), write-only property keys must be assigned
  values that can be fully resolved at the time their values are assigned.
- Property keys may be re-assigned.  Property keys take on their last assigned value.  For
  instance, the following will display "Hello", then display "World":
    test=Hello
    %log=${test}
    test=World
    %log=${test}
- Property key assignments can be conditional if the property key has not already been
  assigned a value by placing a "?" after the key specification.  For instance, in the 
  property assignment "KEY?=VALUE", the property "KEY" will only be assigned "VALUE" if 
  "KEY" has not already been assigned a value previously.
    
The following are reserved read-only constant property keys:
   %version        returns the current version, as in "2.3.4"
   %contextName    returns the context 'name' (name of the servlet context, or main class name)
   %contextPath    returns the context 'path' (path to the servlet context, or $GTS_HOME)
   %configURL      returns the URL for the loaded config file
   %hostName       returns the current host-name
   %hostIP         returns the current mail host IP address

The following are reserved write-only (assignable) constant property keys:
   %log            displays the specified value to the log output
   %include        includes the config at the value URL (URL must exist, error otherwise)
   %include?       includes the config at the value URL (URL may exist, otherwise ignored)
   (The included URL protocol must be one of "file" or "http".)
   
The following are available within a PrivateLabel session context:
   session.name    The PrivateLabel Domain name
   session.locale  The PrivateLabel Domain locale

-----------------------------------------------------------------------------------

All trade names listed above are trademarks of their respective companies.

CHANGELOG.txt
-----------------------------------------------------------------------------------
GTS Changelog
Copyright 2007-2017 GeoTelematic Solutions, Inc.
-----------------------------------------------------------------------------------

Version 2.6.5
- NEW: Added additional status codes.
- NEW: Added "maximumRpm" field to Device table
- UPD: Added support for table Insert/Update "IGNORE" keyword.
- UPD: Added support for table alternate key sort order.
- DCS: Enabled additional support for TKNano, and OBD data.
- UPD: Added partial support for OpenSeaMaps to OpenLayers map support.
- UPD: Updated UnwiredLabs cell-tower lookup to allow including the deviceID.

Version 2.6.4 [2017/04/13]
- NEW: Added additional status codes.
- NEW: Added additional column types to report field layout.
- NEW: Added additional fields to the report FieldLayout and EventDataLayout.
- NEW: Added Account-level SMTP properties.
- NEW: Added additional fields to Device table.
- NEW: Added additional fields to Driver table.
- NEW: Added additional fields available for info-balloon display.
- NEW: Added "setGroupByFields" to DBSelect.
- NEW: Feature to display table records that do not have a valid Account owner.
- NEW: Added optimizations to DailySummaryReport.
- NEW: Added support for loading Pushpins from a directory.
- FIX: Fixed path color specified in the Google static map support.
- UPD: Update map link Google KML to use updated pushpin location.
- DCS: Added "batteryVolts", "vBatteryVolts" field support to GPSEvent.java
- FIX: Fixed DBRecordKey _getWhereClause when no keys specified and partial=all
- DCS: tk10x] Fixed parsing check for VJoy devices (thanks to Xavier Descamps)
- CHG: EventData "emailProperties" renamed to "smtpProperties"

Version 2.6.3 [2016/10/03]
- UPD: EventData report layout now supports "fuelLevelVolume" to include tank #1/#2
- UPD: Default "trackMap.includeStatusCodeColor" to true for Location Detail list.
- UPD: Added additional support to EventData to handle 2 fuel tanks
- NEW: Added option to check for previous passwords when changing passwords.
- NEW: Added "useSSL" option to OpenLayers map provider configuration.
- FIX: Fixed "fuelLevelVolume" in EventData detail/summary reports.
- NEW: "Change Password" page displayed if password expired on login.
- NEW: Updated top menu to automatically adjust displayed menu buttons per row.
- NEW: Added support for Leaflet map tile display engine.
- NEW: Added support for OpenStreetMap using the Leaflet map tile display engine.
- NEW: Added support for ESRI maps using the Leaflet map tile display engine.
- NEW: Added support for MapBox maps using the Leaflet map tile display engine.

Version 2.6.2 [2016/06/10]
- UPD: Updated Generics to support compiling with Java-7 without warning/errors.
- NEW: Can now compile with either Tomcat-7 or Tomcat-8 without code mods.
- NEW: Checkinstall now displays current Tomcat version in use.
- NEW: Added additional status codes.
- NEW: Added summary totals line to "Trip Report Detail".
- NEW: Added support to Device.java for fuel tank #2 capacity.
- NEW: Added "cargoTemp", "coolantPressure" columns to EventData table.
- NEW: Added "sequence" column to GroupList table (to retain list order).
- NEW: Added report total column to TripReport.
- NEW: Added excessive failed login checks to Account/User login.
- NEW: Added additional EventDataLayout/FieldLayout report column support.
- UPD: Changed EventData table "sequence" column to unsigned-long.
- DCS: astra] Update Astra DCS from Astra Telematics Limited.
- FIX: Fixed "addtnlSelect_2" issue in EventData "getRangeEvents" to fix reports.
- FIX: Added fix to detect stale/closed connections for MariaDB compatibility
- FIX: Updated device last-ignition-state to check last ignition on/off times.
- FIX: Fixed "gtsdmtp" compile issue.

Version 2.6.1 [2016/01/17]
- NEW: Added additional status codes.
- NEW: Added additional fields to Device table.
- NEW: Added option to "checkInstall" to display list of modified files.
- NEW: Added "useSSL" option to Google reverse-geocoding configuration.
- NEW: Added support for OpenCageData reverse-geocoding.
- NEW: Map pushpin replay can be now configured to skip over nearby pushpins.
- NEW: Added "tripElapse" column to Driving/Stopped report
- NEW: Added support for fuel-tank #2 fuel level profiling
- NEW: Added additional geozone load/import options
- UPD: "startupInit.Device.OpenDMTPFieldInfo=true" needed to for OpenDMTP fields.
- FIX: Fixed index map link display for Event Detail (by Group) report
- DCS: astra] Added missing "estimateOdometer" support.
- DCS: tk10x] Added support for configuring hearbeat ACK.
- DCS: tk10x] Added "speedConversionFactor" property for custom speed units

Version 2.6.0 [2015/09/24]
- NEW: Added additional status codes.
- NEW: Requires at least Java JDK 7 to utilize additional features.
- NEW: Added "engineTemp", "oilCoolerInTemp", "oilCoolerOutTemp" to EventData.

war/track/js/TrackMap.js
// ----------------------------------------------------------------------------
// Copyright 2007-2017, GeoTelematic Solutions, Inc.
// All rights reserved
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ----------------------------------------------------------------------------
// Change History:
//  2008/08/17  Martin D. Flynn
//     -Moved from TrackMap.java
//  2008/09/19  Martin D. Flynn
//     -Added support for updating map with "Last" point.
//     -Added support for starting "AutoUpdate" on map load.
//  2009/04/02  Martin D. Flynn
//     -Added 'trackMapEventLimit' to allow event 'limit' overrides.
//  2009/11/01  Martin D. Flynn
//     -If "AutoUpdateOnLoad" is true, clicking "Update" now just resets/restarts
//      the auto-update timer.
//  2009/11/10  Martin D. Flynn
//     -Added condition to only restart auto-update if "AutoUpdateOnLoad" is true
//      AND auto-update is currently active/in-process.
// ----------------------------------------------------------------------------

var ID_DEVICE_ID        = "deviceSelector";
var ID_DEVICE_DESCR     = "deviceDescription";
var ID_GOTO_ADDRESS     = "gotoAddressFld";

var PAGE_ZONEGEOCODE    = "ZONEGEOCODE";
var TAG_geocode         = "geocode";
var TAG_geonames        = "geonames";
var TAG_code            = "code";
var TAG_geoname         = "geoname";
var TAG_lat             = "lat";
var TAG_lng             = "lng";
var TAG_lon             = "lon";

// ----------------------------------------------------------------------------

var trackMapLimitOverride = -1;

/* return the event 'limit' for non-fleet maps */
function trackMapEventLimit()
{
    return IS_FLEET? -1 : trackMapLimitOverride;
};

// ----------------------------------------------------------------------------

/* this is executed when the page is loaded */
function trackMapOnLoad()
{

    /* display calendars */
    if (mapCal_fr) {
        mapCal_fr.setCollapsible(CalendarCollapsible, CalendarFade, CalendarDivBox);
        mapCal_to.setCollapsible(CalendarCollapsible, CalendarFade, CalendarDivBox);
        calWriteCalendars(mapCal_fr, mapCal_to);
        if ((CalendarDateOnLoad == "last") && jsvLastEventYMD && (jsvLastEventEpoch > 0)) {
            mapCal_to.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, jsvLastEventYMD.DD);
            mapCal_fr.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, jsvLastEventYMD.DD);
        }
    } else {
        mapCal_to.writeCalendar();
        if ((CalendarDateOnLoad == "last") && jsvTodayYMD) {
            mapCal_to.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, jsvTodayYMD.DD);
        }
    }

    /* init AutoUpdate button text */
    var btn = document.getElementById(ID_MAP_AUTOUPDATE_BTN);
    if (btn != null) { btn.value = AutoUpdateEnable? TEXT_autoUpdateStart : "??"; }

    /* collapse map controls */
    if (ControlCollapseOnLoad) {
        jsmControlDisplay(false);
    }

    /* update map */
    if (AutoUpdateOnLoad) {
        // start AutoUpdate (first RecenterMode should be RECENTER_ZOOM)
        var savedRecenterMode = AutoUpdateRecenterMode;
        AutoUpdateRecenterMode = jsmRecenterZoomMode(RECENTER_ZOOM);
        //trackMapClickedAutoUpdate();
        startAutoUpdateMapTimer();
        AutoUpdateRecenterMode = savedRecenterMode;
    } else {
        // update map points
        var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit();
        trackMapUpdateMap(limit, trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), 0);
    }

};

/* this is executed when the page is unloaded */
function trackMapOnUnload()
{
    mapProviderUnload();
};

// ----------------------------------------------------------------------------

/* this is executed when a new region is selected */
function trackMapSelectRegion()
{

    /* remove current shapes */
    jsmDrawShape("!");

    /* find zoomRegion */
    var zoomRegion = null;
    if (trackZoomRegionShapes) {
        var zrName = document.SelectDeviceForm.regionSelector.value;
        if ((zrName != "") && (zrName != "normal")) {
            for (var i = 0; i < trackZoomRegionShapes.length; i++) {
                if (zrName == trackZoomRegionShapes[i].name) {
                    zoomRegion = trackZoomRegionShapes[i];
                    break;
                }
            }
        }
    }
    if (zoomRegion == null) {
        // selected zoom region not found
        jsmSetFixedZoom(false);
        trackMapClickedUpdateAll();
        return;
    }
    
    /* attributes */
    var color = zoomRegion.color;
    var type  = zoomRegion.type;   // numParseFloat(fld[0], 0);
    var radM  = zoomRegion.radius; // numParseFloat(fld[1], 0);

    // ZoomRegion value:
    // <type>,<radius>,<lat>/<lon>,<lat>/<lon>
    var fld = zoomRegion.points.split(',');
    //if (fld.length < 3) { return; }
    //var type  = numParseFloat(fld[0], 0);
    //var radM  = numParseFloat(fld[1], 0);
    var jspt  = [];
    for (var i = 0; i < fld.length; i++) {
        var LL = fld[i].split('/');
        if (LL.length < 2) { continue; }
        var lat = numParseFloat(LL[0], 0);
        var lon = numParseFloat(LL[1], 0);
        if (((lat != 0) || (lon != 0))) {
            jspt.push(new JSMapPoint(lat,lon));
        }
    }
    if (jspt.length == 0) {
        jsmSetFixedZoom(false);
        trackMapClickedUpdateAll();
        return;
    }
    
    /* legacy zoom mode */
    if ((type == "zoom") || (type == "!zoom")) {
        jsmSetCenter(jspt[0].lat, jspt[0].lon, radM);
        jsmSetFixedZoom(true);
        return;
    }

    /* draw shape */
    if (jsmDrawShape(type,radM,jspt,color,true,"",-1)) {
        jsmSetFixedZoom(true);
        return;
    }
    
    /* not supported */
    jsmSetFixedZoom(false);

};

// ----------------------------------------------------------------------------

/* reset the calendar dates to the last known device date */
function _resetCalandarDates()
{
    if (!IS_FLEET && jsvLastEventYMD && (jsvLastEventEpoch > 0)) {
        if (mapCal_to) {
            var day = jsvLastEventYMD.DD + 1;  // next day
            mapCal_to.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, day);
        }
        if (mapCal_fr) { 
            var day = jsvLastEventYMD.DD;  // this day
            mapCal_fr.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, day); 
        }
    } else
    if (jsvTodayYMD) {
        if (mapCal_to) {
            var day = jsvTodayYMD.DD + 1;  // next day
            //alert("_resetCalandarDates: YYYY="+jsvTodayYMD.YYYY+ " MM="+jsvTodayYMD.MM+ " DD="+day);
            mapCal_to.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, day);
        }
        if (mapCal_fr) { 
            var day = jsvTodayYMD.DD;  // this day
            mapCal_fr.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, day); 
        }
    } else {
        // as a last resort, we could rely on the 'Date' returned by the browser
        // (which may be inaccurate), but for now just leave the calendars as-is.
    }
};

/* this is executed when the "Latest Event" is selected */
function trackMapGotoLastEventDate()
{
    _resetCalandarDates();
    trackMapClickedUpdateAll();
};

// ----------------------------------------------------------------------------

/* this is executed when the "Update Map"/"Auto Update" is pressed */
function trackMapPingDevice(devcmd) 
{
    if (!IS_FLEET && mapDevicePing && DEVICE_PING_URL && (DEVICE_PING_URL != "")) { 
        var dev = document.SelectDeviceForm.device.value; // PARM_DEVICE
        var rfr = mapCal_fr.getArgDateTime();
        var rto = mapCal_to.getArgDateTime();
        var tmz = calGetTimeZone();
        var url = DEVICE_PING_URL + 
            "&_uniq=" + Math.random() +  // necessary to make the URL unique
            "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +        // necessary?
            "&" + PARM_RANGE_TO + "=" + strEncode(rto) +        // necessary?
            "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +        // necessary?
            "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev) +    
            "&" + PARM_DEVICE_COMMAND + "=" + strEncode(devcmd?devcmd:"");
        mapDevicePing(url); // valid iff 'jsmap.js' is in use!
    }
};

// ----------------------------------------------------------------------------

/* get limit type first/last */
function trackMapLimitType()
{
    return LimitType;
};

/* this is executed when the "Update All" is pressed */
function trackMapClickedUpdateAll() // all
{
    if (AutoUpdateOnLoad && (AutoUpdateMapTimer != null)) {
        startAutoUpdateMapTimer();
    } else {
        stopAutoUpdateMapTimer();
        trackMapUpdateMap(trackMapEventLimit(), trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), 0);
    }
};

/* this is executed when the "Update Last" is pressed */
function trackMapClickedUpdateLast() // last
{
    stopAutoUpdateMapTimer();
    trackMapUpdateMap(1, "last", jsmRecenterZoomMode(RECENTER_ZOOM), 0);
};

/* this is executed when the "Replay Map" is pressed */
function trackMapClickedReplay(showInfoBox) 
{
    //try { document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked = false; } catch (e) {}
    stopAutoUpdateMapTimer();
    var replay = showInfoBox? 2 : 1; // 1=don't show info box, 2=show info box (if enabled)
    var replayState = mapProviderPauseReplay(replay);
    if (replayState > 0) {
        // replay paused/resumed
    } else {
        // (re)starting replay
        trackMapUpdateMap(trackMapEventLimit(), trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), replay);
    }
};

/* this is executed when "Update Map" is clicked */
function trackMapUpdateMap(limit, limitType, recenterMode, replay) 
{
    var limitFirst = false;
    //try { document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked = false; } catch (e) {}
    var dev = IS_FLEET? document.SelectDeviceForm.group.value : document.SelectDeviceForm.device.value; // PARM_GEOUP/PARM_DEVICE
    var rfr = mapCal_fr? mapCal_fr.getArgDateTime() : "";
    var rto = mapCal_to.getArgDateTime();
    var tmz = calGetTimeZone();
    var url = MAP_UPDATE_URL + 
        "&_uniq=" + Math.random() +  // necessary to make the URL unique
        "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +
        "&" + PARM_RANGE_TO + "=" + strEncode(rto) +
        "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +
        "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev);
    if (limit > 0) {
        url += "&" + PARM_LIMIT + "=" + limit;
    }
    if (limitType != "") {
        url += "&" + PARM_LIMIT_TYPE + "=" + limitType;
    }
    mapProviderUpdateMap(url, recenterMode, replay);
    // "Replay" button ID is ID_MAP_REPLAY_BTN
};

/* this is executed when "Update KML" is clicked */
function trackMapUpdateKML() 
{
    if (KML_UPDATE_URL && (KML_UPDATE_URL != "")) {
        var dev = IS_FLEET? document.SelectDeviceForm.group.value : document.SelectDeviceForm.device.value; // PARM_GEOUP/PARM_DEVICE
        var rfr = mapCal_fr? mapCal_fr.getArgDateTime() : "";
        var rto = mapCal_to.getArgDateTime();
        var tmz = calGetTimeZone();
        var url = KML_UPDATE_URL +  // Google KML
            "&_uniq=" + Math.random() +  // necessary to make the URL unique
            "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +
            "&" + PARM_RANGE_TO + "=" + strEncode(rto) +
            "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +
            "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev);
        openURL(url, "_blank");
    }
};

/* this is executed when the "Last Location" is pressed */
//function trackMapClickedCenterOnLast()
//{
//    stopAutoUpdateMapTimer();
//    try { jsmCenterOnLastPushpin(document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked); } catch (e) {/*ignore*/}
//};

// ----------------------------------------------------------------------------

/* auto-update map (toggle) */
function trackMapClickedAutoUpdate()
{
    if (AutoUpdateMapTimer != null) {
        stopAutoUpdateMapTimer();
    } else {
        startAutoUpdateMapTimer();
    }
};

/* sets the AutoUpdateMap button text */
var _autoUpdateButtonElem = null;
function _setAutoUpdateButtonText(text)
{
    if (_autoUpdateButtonElem == null) {
        _autoUpdateButtonElem = document.getElementById(ID_MAP_AUTOUPDATE_BTN);
    }
    if (_autoUpdateButtonElem != null) { 
        _autoUpdateButtonElem.value = text; 
    }
};

/* periodic map update timer target */
function _timerAutoUpdateMap() 
{
    if (--AutoIntervalCount <= 0) {
        _resetCalandarDates();
        var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit();
        trackMapUpdateMap(limit, "last", jsmRecenterZoomMode(AutoUpdateRecenterMode), 0);
        if ((AutoMaxCount > 0) && (++AutoUpdateMapCount >= AutoMaxCount)) {
            // we've reached the maximum number of allowed updates.
            stopAutoUpdateMapTimer();
            return;
        }
        AutoIntervalCount = AutoInterval; // start over
    }
    _setAutoUpdateButtonText(TEXT_autoUpdateStop + ' : ' + AutoIntervalCount);
};

/* start a map auto-update timer */
function startAutoUpdateMapTimer() 
{
    stopAutoUpdateMapTimer();
    _resetCalandarDates();
    var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit(); // single last point, or all points
    trackMapUpdateMap(limit, "last", jsmRecenterZoomMode(AutoUpdateRecenterMode), 0); // update map now
    AutoIntervalCount  = AutoInterval;
    AutoUpdateMapCount = 0;
    _setAutoUpdateButtonText(TEXT_autoUpdateStop);
    AutoUpdateMapTimer = setInterval('_timerAutoUpdateMap()',1000); // setTimeout
};

/* stop any running map auto-update timer */
function stopAutoUpdateMapTimer()
{
    if (AutoUpdateMapTimer != null) {
        clearInterval(AutoUpdateMapTimer); // clearTimeout
        AutoUpdateMapTimer = null;
    }
    AutoIntervalCount  = 0;
    AutoUpdateMapCount = 0;
    _setAutoUpdateButtonText(TEXT_autoUpdateStart);
};

// ----------------------------------------------------------------------------

/* show device selector */
// only valid if DeviceChooser has been included
function trackMapShowSelector()
{
    if (deviceShowChooserList) {
        var list = (typeof TrackSelectorList != 'undefined')? TrackSelectorList : null;
        deviceShowChooserList(ID_DEVICE_ID, ID_DEVICE_DESCR, list); 
    }
};

function deviceDeviceChanged()
{
    trackMapSelectDevice();
};

/* this is executed when a new device is selected */
function trackMapSelectDevice()
{ // 'PARM_PAGE=map' && PARM_DEVICE are defined in FORM_NEW_DEVICE
    document.SelectDeviceForm.date_fr.value = mapCal_fr? mapCal_fr.getArgDateTime() : ""; // PARM_RANGE_FR
    document.SelectDeviceForm.date_to.value = mapCal_to.getArgDateTime(); // PARM_RANGE_TO
    document.SelectDeviceForm.date_tz.value = calGetTimeZone(); // PARM_TIMEZONE
    document.SelectDeviceForm.submit();
};

// ----------------------------------------------------------------------------

function trackMapGotoAddress()
{
    var addrFld = document.getElementById(ID_GOTO_ADDRESS);
    if (addrFld == null) { return; }
    var addr = addrFld.value;
    var ct   = "US";

    /* get the latitude/longitude for the zip */
    var url = "./Track?page=" + PAGE_ZONEGEOCODE + "&addr=" + addr + "&country=" + ct + "&_uniq=" + Math.random();
    //alert("URL " + url);
    try {
        var req = jsmGetXMLHttpRequest();
        if (req) {
            req.open("GET", url, true);
            //req.setRequestHeader("CACHE-CONTROL", "NO-CACHE");
            //req.setRequestHeader("PRAGMA", "NO-CACHE");
            //req.setRequestHeader("If-Modified-Since", "Sat, 1 Jan 2000 00:00:00 GMT");
            req.onreadystatechange = function() {
                if (req.readyState == 4) {
                    var lat = 0.0;
                    var lon = 0.0;
                    for (;;) {
                        
                        /* get xml */
                        var xmlStr = req.responseText;
                        if (!xmlStr || (xmlStr == "")) {
                            break;
                        }
                        
                        /* get XML doc */
                        var xmlDoc = createXMLDocument(xmlStr);
                        if (xmlDoc == null) {
                            break;
                        }

                        /* try parsing as "geocode" encasulated XML */
                        var geocode = xmlDoc.getElementsByTagName(TAG_geocode);
                        if ((geocode != null) && (geocode.length > 0)) {
                            //alert("geocode: " + xmlStr);
                            var geocodeElem = geocode[0];
                            if (geocodeElem != null) {
                                var latn = geocodeElem.getElementsByTagName(TAG_lat);
                                var lonn = geocodeElem.getElementsByTagName(TAG_lng);
                                if (!lonn || (lonn.length == 0)) { lonn = geocodeElem.getElementsByTagName(TAG_lon); }
                                if ((latn.length > 0) && (lonn.length > 0)) {
                                    lat = numParseFloat(latn[0].childNodes[0].nodeValue,0.0);
                                    lon = numParseFloat(lonn[0].childNodes[0].nodeValue,0.0);
                                    break;
                                }
                            }
                            break;
                        }

                        /* try parsing as forwarded XML from Geonames */
                        var geonames = xmlDoc.getElementsByTagName(TAG_geonames);
                        if ((geonames != null) && (geonames.length > 0)) {
                            //alert("geonames: " + xmlStr);
                            // returned XML was forwarded as-is from Geonames
                            var geonamesElem = geonames[0];
                            var codeList = null;
                            if (geonamesElem != null) {
                                codeList = geonamesElem.getElementsByTagName(TAG_code);
                                if (!codeList || (codeList.length == 0)) {
                                    codeList = geonamesElem.getElementsByTagName(TAG_geoname);
                                }
                            }
                            if (codeList != null) {
                                for (var i = 0; i < codeList.length; i++) {
                                    var code = codeList[i];
                                    var latn = code.getElementsByTagName(TAG_lat);
                                    var lonn = code.getElementsByTagName(TAG_lng);
                                    if ((latn.length > 0) && (lonn.length > 0)) {
                                        lat = numParseFloat(latn[0].childNodes[0].nodeValue,0.0);
                                        lon = numParseFloat(lonn[0].childNodes[0].nodeValue,0.0);
                                        break;
                                    }
                                }
                            }
                            break;
                        }
                        
                        /* break */
                        //alert("unknown: " + xmlStr);
                        break;
                        
                    }

                    /* set lat/lon */
                    if ((lat != 0.0) || (lon != 0.0)) {
                        // got lat/lon
                        //alert("Lat="+lat +", Lon="+lon);
                        //jsmSetCenter(lat,lon,PUSHPIN_ZOOM);
                        jsmDrawShape("!");
                        jsmDrawShape("circle",200,[ new JSMapPoint(lat,lon) ],"#888855",true,"",-1);
                    }

                } else
                if (req.readyState == 1) {
                    // alert('Loading GeoNames from URL: [' + req.readyState + ']\n' + url);
                } else {
                    // alert('Problem loading URL? [' + req.readyState + ']\n' + url);
                }
            }
            req.send(null);
        } else {
            alert("Error [trackMapGotoAddress]:\n" + url);
        }
    } catch (e) {
        alert("Error [trackMapGotoAddress]:\n" + e);
    }

};

// ----------------------------------------------------------------------------

/* parse geozones onLoad */
function _zoneMapOnLoad() {
    trackMapOnLoad();
};

/* unload */
function _zoneMapOnUnload() {
    trackMapOnUnload()
};

// ----------------------------------------------------------------------------
